### ISSUE TITLE
Add a greedy "list argument"

### ISSUE THREAD

**Issue by JorelAli** [JorelAli/CommandAPI#282]
From #275.

**Please describe your suggestion**
Allowing users to enter a list of values is very useful. WorldEdit has a nice implementation of a list argument, where users can provide a list of materials to manipulate blocks:
![worldedit](https://user-images.githubusercontent.com/4613171/167111609-fced4020-a37a-4860-b814-10fe754c60dd.gif)

Implementing this "properly" is outside of the basic scope of the CommandAPI for now (see [regex argument mod](https://github.com/JorelAli/MinecraftRegexArgumentMod#usage-examples)), however we can "simulate" a list argument by using an underlying `GreedyStringArgument` with a powerful suggestion provider.

Work for this has already begun on the [dev/listargument](https://github.com/JorelAli/CommandAPI/tree/dev/listargument) branch.

Formally, we need to support:
- Providing a list of values for the list
- Ensuring that the inputs to the argument don't accept values outside of the provided list (e.g. a list `[apple, banana, cat]` can't (or won't) accept `dirt` as a value)
- Allowing users to provide tooltips to entries
- Allowing users to specify a delimiter (e.g. a comma vs a space)
- The casting type should be a `Collection` parameterized over some type `T` which is provided by the users

**Please supply examples of the desired inputs and outputs**
_Example:_
Input: Using the following code:

```java
List<String> list = List.of("apple", "banana", "cat", "dog", "elephant");
new CommandAPICommand("blah")
    .withArguments(new ListArgumentBuilder<String>("node", " ").withList(list).withStringMapper().build())
    .executes((sender, args) -> {
        Collection<String> myList = (Collection<String>) args[0];
        sender.sendMessage("Provided: " + myList);
    })
    .register();

new CommandAPICommand("blahcomma")
.withArguments(new ListArgumentBuilder<String>("node", ",").withList(list).withStringMapper().build())
.executes((sender, args) -> {
    Collection<String> myList = (Collection<String>) args[0];
    sender.sendMessage("Provided: " + myList);
})
.register();
```

Output: Typing a command can accept a list of arguments:

![listargument](https://user-images.githubusercontent.com/4613171/167112153-064cfbc7-c303-4544-8493-45165d4fc13f.gif)

**Comment by SNWCreations** [comment]
wow, that is great.
But I think it should be like a single parameter, I mean, each parameter has a separate suggestion box instead of multiple parameters linked together to form a suggestion
For example, when aa bb is suggested for the next parameter, it should only display reasonable suggestions, instead of bringing the previous ones.

**Comment by JorelAli** [comment]
> wow, that is great. But I think it should be like a single parameter, I mean, each parameter has a separate suggestion box instead of multiple parameters linked together to form a suggestion For example, when aa bb is suggested for the next parameter, it should only display reasonable suggestions, instead of bringing the previous ones.

We can't do that. If you do that, if you press TAB for tab-completion, it deletes the current list:
![deletedlist](https://user-images.githubusercontent.com/4613171/167295032-e203dadd-973d-4e4c-9857-d6b8747794a6.gif)

**Comment by SNWCreations** [comment]
> > wow, that is great. But I think it should be like a single parameter, I mean, each parameter has a separate suggestion box instead of multiple parameters linked together to form a suggestion For example, when aa bb is suggested for the next parameter, it should only display reasonable suggestions, instead of bringing the previous ones.
> 
> @SNWCreations We can't do that. If you do that, if you press TAB for tab-completion, it deletes the current list: ![deletedlist](https://user-images.githubusercontent.com/4613171/167295032-e203dadd-973d-4e4c-9857-d6b8747794a6.gif)

Okay... I think you can do this is very good, thank you for your efforts.

**Comment by SNWCreations** [comment]
I'm looking forward to this update. :)
👍

**Comment by SNWCreations** [comment]
I have a question.
Can we check the argument before executing?

**Comment by JorelAli** [comment]
Implemented in 8.2.0.

@SNWCreations Yes, dynamic lists are supported, you can read all about it in the 8.2.0 documentation for list arguments here: https://commandapi.jorel.dev/8.2.0/listarguments.html

**Comment by SNWCreations** [comment]
> Implemented in 8.2.0.
> 
> @SNWCreations Yes, dynamic lists are supported, you can read all about it in the 8.2.0 documentation for list arguments here: https://commandapi.jorel.dev/8.2.0/listarguments.html

Thank you! I will upgrade my project in future. (I am too busy now.)
